independent-projects/arc/runtime/src/main/java/io/quarkus/arc/RequestContext.java

```diff
@@ -101,32 +86,34 @@ public void destroy(Contextual<?> contextual) {
         }
     }
 
-    private void fireIfNotEmpty(LazyValue<EventImpl.Notifier<Object>> value) {
-        EventImpl.Notifier<Object> notifier = value.get();
-        if (!notifier.isEmpty()) {
-            notifier.notify(toString());
+    @Override
+    public void activate(ContextState initialState) {
+        if (initialState == null) {
+            currentContext.set(new ConcurrentHashMap<>());
+        } else {
+            if (initialState instanceof RequestContextState) {
+                currentContext.set(((RequestContextState) initialState).value);
+            } else {
+                throw new IllegalArgumentException("Invalid inital state: " + initialState);
+            }
         }
+        // Fire an event with qualifier @Initialized(RequestScoped.class) if there are any observers for it
+        fireIfNotEmpty(initializedNotifier);
     }
 
     @Override
-    public void deactivate() {
-        currentContext.remove();
+    public InjectableContext.ContextState getState() {
+        ConcurrentMap<Contextual<?>, ContextInstanceHandle<?>> ctx = currentContext.get();
+        if (ctx == null) {
+            // Thread local not set - context is not active!
+            throw new ContextNotActiveException();
```

Comment:
I'm not sure this is the best way to do this. For users it would be easier to have a `null` state or a singleton marker state that could be used in `activate` to call `currentContext.remove()`. Also you currently accept `null` in `activate` but it doesn't do `deactivate` which is confusing.

Otherwise callers will have to capture/restore the state by looking first at whether the request scope is active or not and that complexifies everything. If we treat inactive request contexts as a special state it makes things simpler.

Reply:
> For users it would be easier to have a null...

I disagree. Throwing a `ContextNotActiveException` is very common in the CDI API (see for example `BeanManager.getContext()`). A user/integrator can always check `Context.isActive()` and Arc has even `ArcContainer.getActiveContext(Class<? extends Annotation>)` which returns the active context or null.

>  Also you currently accept null in activate but it doesn't do deactivate which is confusing.

I don't understand.

Reply:
Well, look at the `ArcContextProvider` class: it needs to deal with deactivated contexts on the capturing side and on the restoring side. If an inactive context was a special `ContextState` it would make the API a lot simpler to use. It's cosmetic, though, we can make the `ArcContextProvider` as complex as we need to if required.

Reply:
> Throwing a ContextNotActiveException is very common in the CDI API

That's true.

> Also you currently accept null in activate but it doesn't do deactivate which is confusing.

Null just means that you make a new backing map, the deactivation is then done as usual.